### ISSUE TITLE
Why the t_correct is larger than t_wrong in MSE loss function

### ISSUE THREAD

**Issue by Xinhu-L** [JulianGoeltz/fastAndDeep#7]
Hi, I noticed that in the default config files, the t_correct is larger than t_wrong, which means that the loss function implies the the right spike comes later( if I understand correctly).
```YAML
  loss:
    type: MSE
    t_correct: 1.7
    t_wrong: 1.0
    # t_correct: 1.0
    # t_wrong: 2.0
```
But I think that the t_correct should be smaller than the t_wrong to train the right spike to come faster(The annotated one). Could you please explain why?

**Comment by JulianGoeltz** [comment]
Hey Xinhu,

good observation. In the MSE case one sets target times, and this slightly changes the way classes are selected: [in the code](https://github.com/JulianGoeltz/fastAndDeep/blob/main/src/utils.py#L449) you can see that the classification of the network is determined by the label spike *closest* to the target spike time.
Approximately, the committed configuration comes down to classification by *the last spike*. A reason for this choice was detailed already in the paper, the chance to integrate more information, but the values are by far not optimised and you can try your configuration with `t_corr < t_wrong` as well.

I hope this helps, if you have further questions feel free to ask.
Cheers,
Julian

**Comment by Xinhu-L** [comment]
Hi Julian,

Thanks a lot for your reply, which is quite clear. So the two loss functions will train the network in two ways: 1. the cross-entropy will train the right label spike to come earlier while 2. the MSE will train the right label spike happens close to t_correct. Right?
And you also said that "A reason for this choice was detailed already in the paper, the chance to integrate more information", which still puzzled me. Your paper mentioned that in TTFS, the earlier spike carries more information, so if you set the t_correct later than t_wrong, it conflicts with the conclusion. 🤔 
Please correct me if I found the wrong conclusion from your paper or if there is still something that I missed. : )

Best,
Xinhu

**Comment by JulianGoeltz** [comment]
It's a pleasure :)
> the cross-entropy will train the right label spike to come earlier

yes, but more precisely it will increase the distance between the correct and wrong times.
> the MSE will train the right label spike happens close to t_correct 

Again correct, but additionally it will train to have the wrong label spikes close to t_wrong.


> And you also said that "A reason for this choice was detailed already in the paper, the chance to integrate more information", which still puzzled me. Your paper mentioned that in TTFS, the earlier spike carries more information, so if you set the t_correct later than t_wrong, it conflicts with the conclusion. :thinking:

I think there's a small confusion: in the TTFS coding as used in the paper, we *interpret* an earlier spike of a neuron with a stronger representation of the feature that neuron encodes. One could say, as you did, that this earlier spike 'carries more information', but in the sense that the feature is more strongly represented.
What I wanted to hint at is that if the label layer reaches a decision later in time, than naturally more input spikes will have happened by that time, therefore the decision can be based on information of more inputs.

Cheers,
Julian

**Comment by Xinhu-L** [comment]
Aha, I see what you mean, by manually setting the t_correct late, more neurons in the last hidden layer(or the layer before the output layer) will spike so that the output layer will have more input spikes, which means "more chance to integrate more information".
While the confusion is like, in the TTFS encoding, the earlier the spike is, the “more important” the spike is, but here in your MSE loss function the later spike carries "more” information. So it's the difference between "more important" and "more", right?
If my explanation above is right, then I think I understand what you mean now, thank u so much for the explanation, which is super clear. 😃
